habitat-hitl/habitat_hitl/environment/controllers/controller_helper.py

```diff
@@ -77,52 +78,113 @@ def __init__(
+                            original_action_space
+                        )
+
+                        (
+                            base_vel_action_idx,
+                            base_vel_action_end_idx,
+                        ) = find_action_range(
+                            original_action_space, "_base_velocity"
+                        )
+
+                        assert len(action_space.shape) == 1
+                        num_actions = action_space.shape[0]
+
+                        articulated_agent = self._env._sim.agents_mgr[agent_index].articulated_agent  # type: ignore[attr-defined]
+
+                        # sloppy: derive turn scale. This is the change in yaw (in radians) corresponding to a base ang vel action of 1.0. See also Habitat-lab BaseVelAction.
+                        turn_scale = (
+                            config.habitat.simulator.ctrl_freq
+                            / config.habitat.task.actions.agent_0_base_velocity.ang_speed
+                        )
+
+                        gui_agent_controller = GuiRobotController(
+                            agent_idx=agent_index,
+                            is_multi_agent=is_multi_agent,
+                            gui_input=gui_input,
+                            articulated_agent=articulated_agent,
+                            num_actions=num_actions,
+                            base_vel_action_idx=base_vel_action_idx,
+                            num_base_vel_actions=base_vel_action_end_idx
+                            - base_vel_action_idx,
+                            turn_scale=turn_scale,
+                        )
+                    else:
+                        raise ValueError(
+                            f"articulated agent type {articulated_agent_type} not supported"
+                        )
+
+                    self.controllers.append(gui_agent_controller)
+
+                else:
+                    self.controllers.append(
+                        SingleAgentBaselinesController(
+                            agent_index,
+                            is_multi_agent,
+                            config,
+                            self._gym_habitat_env,
+                        )
+                    )
 
-    def get_gui_controlled_agent_index(self) -> Optional[int]:
-        return self._gui_controlled_agent_index
+    def _find_gui_controlled_agent_config(self, agent_index):
+        for (
+            gui_controlled_agent_config
+        ) in self._hitl_config.gui_controlled_agents:
+            if gui_controlled_agent_config.agent_index == agent_index:
+                return gui_controlled_agent_config
+        return None
+
+    def get_gui_agent_controllers(self) -> List[Controller]:
+        """
+        Return list of controllers indexed by user index. Beware the difference between user index and agent index. For example, user 0 may control agent 1.
+        """
```

Comment:
>agent index concept should be entirely abstracted away at the HITL level
I doubt this will be feasible. We shouldn't think of the controller as completely encapsulating the agent. It is really just encapsulating how we give actions to the agent. E.g. if you want to know something particular about the agent state (e.g. arm position), you would probably still use the Habitat-lab agent API including knowing the particular agent index.